inc/Engine/CriticalPath/RESTGenerate.php

```diff
@@ -0,0 +1,302 @@
+	 *
+	 * @since 3.6
+	 *
+	 * @param  WP_REST_Request $request WP REST request response.
+	 * @return WP_REST_Response
+	 */
+	public function generate( WP_REST_Request $request ) {
+		$status = get_post_status( $request['id'] );
+
+		if ( ! $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_exists',
+					'message' => __( 'Requested post does not exist.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		if ( 'publish' !== $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_published',
+					'message' => __( 'Cannot generate CPCSS for unpublished post.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		$post_url  = get_permalink( $request['id'] );
+		$post_type = get_post_type( $request['id'] );
+
+		$response = $this->send_generation_request( $post_url, $post_type );
+
+		if ( false === $response['success'] ) {
+			return rest_ensure_response( $response );
+		}
+
+		while ( $job_data = $this->get_critical_path( $response['data']['id'] ) ) { // phpcs:ignore WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition
+			if ( in_array( (int) $job_data->status, [ 400, 404 ], true ) ) {
+				// translators: %1$s = post URL.
+				$error = sprintf( __( 'Critical CSS for %1$s not generated.', 'rocket' ), $post_url );
+
+				if ( isset( $job_data->message ) ) {
+					// translators: %1$s = error message.
+					$error .= ' ' . sprintf( __( 'Error: %1$s', 'rocket' ), $job_data->message );
+				}
+
+				return rest_ensure_response(
+					[
+						'success' => false,
+						'code'    => 'cpcss_generation_failed',
+						'message' => $error,
+						'data'    => [
+							'status' => 400,
+						],
+					]
+				);
+			}
+
+			if ( isset( $job_data->data->state, $job_data->data->critical_path ) && 'complete' === $job_data->data->state ) {
+				if ( ! $this->save_post_cpcss( $request['id'], $post_type, $job_data->data->critical_path ) ) {
+					return rest_ensure_response(
+						[
+							'success' => false,
+							'code'    => 'cpcss_generation_failed',
+							'message' => sprintf(
+								// translators: %1$s = post URL, %2$s = error message.
+								__( 'Critical CSS for %1$s not generated. Error: %2$s', 'rocket' ),
+								$post_url,
+								// translators: %s = critical CSS directory path.
+								sprintf( __( 'The critical CSS content could not be saved as a file in %s.', 'rocket' ), $this->critical_css_path )
+							),
+							'data'    => [
+								'status' => 400,
+							],
+						]
+					);
+				}
+
+				return rest_ensure_response(
+					[
+						'success' => true,
+						'code'    => 'cpcss_generation_successful',
+						// translators: %s = post URL.
+						'message' => sprintf( __( 'Critical CSS for %s generated.', 'rocket' ), $post_url ),
+						'data'    => [
+							'status' => 200,
+						],
+					]
+				);
+			}
+
+			sleep( 2 );
```

Comment:
Probably to prevent DDOSing the API.

Reply:
Yes, it' a pause before sending a new request to the API endpoint and see if the job is complete. There might be a better way to do that, I took this part from our existing code for the CPCSS generation for generic types.

Reply:
The pause is not an indicator that the job is complete or that it's okay for a new API endpoint request. All it does is pause.

So yes, see if there's another way to get a positive indication for processing another request.

Reply:
Details of the API are here: https://github.com/wp-media/criticalpath-css-generator/wiki/API.

Currently it doesn't seem there is any way to know that the job is complete other than doing requests to the endpoint again, until the status is `complete` in the JSON response.

The 2 seconds pause was determined when we discussed the initial implementation of the CPCSS feature.

I believe we would need an API update to handle that better?

Reply:
Okay. Interesting.

One thing you might not know is this: `sleep()` stops all of the PHP execution. That's bad for performance as it blocks other things from running.

A way to handle this could be:

- Add the requests to the background handler.
- Add a acknowledge handshake back from the API to let us know that it's done or an error happened.

What do you think @Tabrisrp?

Reply:
I'm trying to figure out the best way to handle this. Currently the process with the API is as follow:

1. WP REST API POST request to the CPCSS generation endpoint for posts
2. Send a POST request to queue the job for the current post URL
3. If successful, send GET requests to the specific job endpoint until the response states it's `complete` or an error happened
4. Send the response from the WP REST API endpoint to be displayed in the UI (via AJAX)

The blocking point is 3. If we put it on the background process, how do we get the result to 4? The REST request won't wait for the background process to be done.

How could we modify the API to improve this? This would also require Kuba to have allocated time to do so

Reply:
@Tabrisrp Can you lay out a workflow diagram for the communication and handling here? Then we can go through together and figure out how to add any necessary handshaking.

Reply:
I added more details to our existing diagram to highlight the CPCSS API part, in yellow: https://drive.google.com/file/d/1K5vPc0_JURHWiWX7KujZAo1OhJczZbpj/view?usp=sharing